### DISCUSSION TITLE
Protected route on client side and redirect before fetch data?

### DISCUSSION THREAD

**Discussion by namlq93** [vercel/next.js#17351] [Help]
Now, I have user role in Context. I can check the role of context in useEffect on each page before fetch data and redirect to login if the user doesn't have the admin role:
```
useEffect(() => {
    if (userInfo.role == 2) {
        get...;
    } else {
        Router.push('/login');
    }
    
}, []);
```

But I don't want to rewrite to check the user role on each page. How can I use it for every page? After research, I've used a HOC component to protect the route and wrap each page:

```
export const withPrivateRoute = Component => props => {
    const role = 1;

    useEffect(() => {
        if (role == 1) Router.push('/login');
    }, []);

    return <Component {...props} />;
};
```

Fetch data on page (don't need check role because check from HOC component): 
```
useEffect(() => {
        get...
}, []);
```

But, It still fetches data before redirect, and I get an error because can't get API without the admin role: 
![Screen Shot 2020-09-25 at 11 08 13](https://user-images.githubusercontent.com/41458989/94225641-c442c280-ff1f-11ea-88dc-6781326c0b07.png)

Is this the correct way? Thank you for your support!

**Comment by lfades** [comment]
I don't think there's enough code to see where the error is. However there are some things I can tell that may be helpful:

- useEffect will run again when you change the page even if it has empty dependencies, because the old page is unmounted and a new page is mounted, with a different useEffect, so if you're doing a query, it will run again (you may want to cache the result in memory and/or use http headers to have the data cached)
- You may be having an unrelated error in your code if you're getting "unexpected end of JSON input"
